File: Algorithm/src/main/java/com/zhangdd/sort/MergeSort.java
package com.zhangdd.sort;

import java.util.Arrays;

/**
 * @author zhangdd on 2021/11/2
 */
public class MergeSort {

    public static void main(String[] args) {
//        int[] arr = {7, 3, 10, 2, 4, 1, 15, 21, 8, 3};
        int[] arr = {7, 3, 4};
        int[] temp = new int[arr.length];
        mergeSort(arr, 0, arr.length - 1, temp);

        System.out.println(Arrays.toString(arr));
    }

    private static void mergeSort(int[] arr, int left, int right, int[] temp) {
        if (left < right) {
            int mid = (left + right) / 2;
            mergeSort(arr, left, mid, temp);
            mergeSort(arr, mid + 1, right, temp);

            merge(arr, left, mid, right, temp);
        }

    }

    private static void merge(int[] arr, int left, int mid, int right, int[] temp) {

        int l = left;
        int r = mid + 1;
        int tempIndex = 0;

        //1.
        //将arr中，left到mid 和mid+1到right两个数组中的数按序填充到temp中
        //直到一个子数组空了
        while (l <= mid && r <= right) {
            if (arr[l] < arr[r]) {
                temp[tempIndex] = arr[l];
                l++;
                tempIndex++;
            } else {
                temp[tempIndex] = arr[r];
                r++;
                tempIndex++;
            }
        }

        //2.
        //将剩余的一个子数组内的数据填充到temp中
        while (l <= mid) {
            temp[tempIndex] = arr[l];
            l++;
            tempIndex++;
        }

        while (r <= right) {
            temp[tempIndex] = arr[r];
            r++;
            tempIndex++;
        }

        tempIndex = 0;
        int tempLeft = left;
        //3.
        //将这次temp中填充的数据拷贝到arr中
        while (tempLeft <= right) {
            arr[tempLeft] = temp[tempIndex];
            tempIndex++;
            tempLeft++;
        }
    }
}

File: python/Algorithm/sort/MergeSort.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/2/4
"""


def merge_sort(arr):
    left_index = 0
    right_index = len(arr) - 1
    do_merge_sort(arr, left_index, right_index)


def do_merge_sort(arr, left_index, right_index):
    if left_index < right_index:
        middle = (left_index + right_index) // 2
        # 对左边做归并排序
        do_merge_sort(arr, left_index, middle)
        # 对右边做归并排序
        do_merge_sort(arr, middle + 1, right_index)

        # 对当前的左右两组进行合并
        merge(arr, left_index, right_index, middle)


def merge(arr, left_index, right_index, middle):
    l = left_index
    r = middle + 1
    temp = []
    while l <= middle and r <= right_index:
        if arr[l] < arr[r]:
            temp.append(arr[l])
            l += 1
        else:
            temp.append(arr[r])
            r += 1

    while l <= middle:
        temp.append(arr[l])
        l += 1

    while r <= right_index:
        temp.append(arr[r])
        r += 1

    temp_left = left_index
    temp_index = 0
    while temp_left <= right_index:
        arr[temp_left] = temp[temp_index]
        temp_index += 1
        temp_left += 1


if __name__ == '__main__':
    collection = [8, 4, 5, 7, 1, 3, 6, 2]
    merge_sort(collection)
    print(collection)
